### ISSUE TITLE
FSD：上下文分支（Context）

### ISSUE THREAD

**Issue by wlvh** [wlvh/RBT_RL#3]
# FSD：上下文分支（Context）

**版本**：context_v1.0
**目标系统**：多分支 + 图时空 + 条件专家（MoE）的监督/排序学习框架
**关注范围**：仅“上下文分支”的**输入/规则/输出**、与其他分支的**交互契约**与**验收标准**。
**不在本 FSD 范围内**：参数分支（strategy params）、结构分支（type/target/time）、生成器实现、训练调度、回测执行。

---

## 1. 概览与目标（Overview & Objective）

**业务目标**
在仅有 BTC、200 周、每周 300 个策略的低样本设置下，**最小化过拟合、最大化样本外稳健性**。上下文分支提供**市场环境表示**，并用于：

1. 作为排序/分类/回归主干的**上下文嵌入**；
2. 向条件专家（MoE）提供**market_regime 门控信号**；
3. 为生成器/门控质检提供**上下文摘要**（可选）。

**设计原则**

* **强归纳偏置优先**：工程特征（稳健 TAs + 统计）为主，辅以**可学习软窗口**汇聚，避免“固定 30 天”的硬编码。
* **零泄漏**：所有窗口均严格**只用当前周（含）之前**的数据。
* **可治理、可回放**：特征版本化、质检指标、数据掩码、分布监控。
* **渐进式可扩展**：可插拔“小型原始序列支路（TCN）/ST-GT 分支”（默认关闭）。

---

## 2. 全局定义与术语表（Glossary）

* **周（week_t）**：策略参数更新与标签评估的时间粒度，t 为周索引。
* **as_of_date**：周 t 的结束日（通常为周日）；**允许使用的数据**为 `as_of_date` 及其之前。
* **market_window(W)**：以 `as_of_date` 向后回看 W 天的市场窗口（不含未来）。
* **窗口集合（W_set）**：默认 `{7, 14, 21, 30, 60}` 天。
* **engineered features（TAs+Stats）**：稳健的技术指标与统计摘要（定义见 §3.2）。
* **soft-window aggregator**：对同一原始特征在不同 W 的派生做**非负权重**汇聚，权重由上下文决定并经 Softmax 归一。
* **mask**：缺失/不可用数据的二值指示；**掩码语义必须传递**到下游模型。
* **market_regime**：市场状态标签（低/中/高波动等），由上下文分支派生，**作为 MoE 门控输入**。
* **purged walk-forward**：滚动训练/验证/测试，邻近周严格 purge，避免信息泄漏。

---

### 3.1 输入（Inputs）

**I-1 数据源**

* **价格与成交量**（日频）：OHLCV，来源为经校验的市场数据表。
* **可选外生上下文**（如资金费率、链上摘要、期权隐波）：若不存在则缺省。
* **时间索引**：`as_of_date`（每个周 t 的结束日）。

**I-2 输入契约**

* 对每个 `as_of_date`，上下文分支接收：

  ```
  {
    "date": as_of_date (datetime),
    "market_slice": {   # 至多回看 max(W_set) 天的原始列
        "daily_ohlcv": DataFrame[date, open, high, low, close, volume],
        "exogenous":  (可选) DataFrame[date, ...]
    },
    "config": {
        "W_set": [7,14,21,30,60] (默认),
        "robust_norm": {"method": "robust|quantile|log", "window": [W_set最大]},
        "winsorize": {"p_low": 0.01, "p_high": 0.99},
        "feature_version": "features_v1",
        "allow_exogenous": true|false
    }
  }
  ```
* **边界条件**：当 `as_of_date` 历史长度不足 `max(W_set)` 时，允许短窗口计算，并对不可得窗口输出 `mask=0`、数值置 0（或 NaN→0）并上报 `coverage_ratio`。

---

#### 3.2.1 特征派生（engineered features，默认特征字典）

**A. 基础价量 TAs**（全部逐列加入 `mask`）

* **收益与动量**：`ret_k`（k=1,5,10,20 天）= `close/close.shift(k)-1`；`mom_k`=累计回报。
* **波动率**：`rv_k`（realized vol，日收益标准差×√年化）；`parkinson_k`（高低波动）。
* **趋势与摆动**：`rsi_k`（Wilder）、`macd(12,26,9)` 的 DIF/DEA、`atr_k/price`。
* **量能**：`vol_chg_k`、量价相关 `corr_k(price, volume)`。
* **价差与回撤**：`maxdrawdown_k`、`amplitude_k = (high-low)/close`。

**B. 稳健统计摘要**（对 close/ret 的滚动统计）

* `mean_k, std_k, skew_k, kurt_k, q10_k, q25_k, q50_k, q75_k, q90_k`。
* 极值比与尾部比率：`max/min 比, (q90 - q10)/q50`。

**C. 外生上下文（可选）**

* 资金费率均值/分位、隐波水平/偏斜（若可得）；若不可得，保持空并 `mask=0`。

> **说明**：上面指标以**多个窗口 W in W_set**同时计算，得到一组**分尺度特征**。所有计算仅使用 `as_of_date` 及之前的日数据。

#### 3.2.2 标准化与稳健处理

* **Winsorize**：按 `p_low=1%/p_high=99%` 分位对每列裁剪（在“训练期”分位上固定）。
* **标准化**：`robust | quantile | log` 三选一（默认 `robust`：减中位数、除 IQR）。
* **缺失处理**：缺失 → 0；并**产生同名 `_mask` 列**（1=可用、0=缺失）。

#### 3.2.3 多尺度软窗口汇聚（soft-window aggregator）

* 对任一原始特征 `f` 及其尺度扩展 `f_w`（w∈W_set），计算：
  [
  \alpha_w = \text{softmax}(g(z))*w,\quad z = \text{context_summary}({f_w},\text{vol/ret摘要})
  ]
  [
  f*{\text{multi}} = \sum_{w\in W_set} \alpha_w \cdot f_w
  ]
* `g` 为轻量 MLP（≤ 2 层，隐藏维 32–64），共享于同类特征；
* 约束：`α_w ≥ 0` 且 `∑ α_w = 1`；**将 α 记录到日志**用于监控。
* 目的：避免**固定 30 天**僵化，保留有限自由度，**降低过拟合**。

#### 3.2.4 市场状态识别（market_regime）

* **默认法（稳健低算力）**：基于 `rv_30`（或 `rv_multi`）的**三分位切分**：

  * `regime ∈ {Low, Mid, High}`，阈值在**训练期固定**，推断期只应用。
* **可选法（升级）**：HMM(3) on `ret_1`/`rv_30` 联合，输出最可能状态；默认关闭。
* 产出：`regime_onehot(3)` 或 `regime_embed(8-d)`。

#### 3.2.5 小型原始序列支路（可选，默认关闭）

* **轻量 TCN**：1–2 层，通道≤16，Kernel≤5，仅处理 `close/volume` 的标准化序列（长度= max(W_set)）；
* 输出经过**强正则**（dropout 0.2，L2 1e-4）后拼接到上下文；
* 仅当 `config.enable_raw_branch=true` 时启用。

---

### 3.3 输出（Outputs）

上下文分支输出对周 t 的**上下文向量**与**门控信号**：

```
{
  "context_vec":  FloatTensor[D_context],  # 默认 D_context = 128
  "market_regime": {
      "regime_id": Int[1],                 # {0,1,2} = {Low,Mid,High}
      "regime_onehot": FloatTensor[3]
  },
  "soft_window_alpha": Dict[str, FloatTensor[len(W_set)]],  # 每类特征的 α_w
  "masks": Dict[str, IntTensor[1]],        # 汇总的可用性（或逐列 _mask）
  "meta": {
      "feature_version": "features_v1",
      "W_set": [7,14,21,30,60],
      "coverage_ratio": Float[0..1],       # 实际可计算窗口比例
      "as_of_date": datetime
  }
}
```

**契约要求**

* `context_vec` 必须**输入**：

  1. **主排序/分类/回归头**（与参数分支、结构分支 concat 或加性融合）；
  2. **MoE 的 gating 模块**（参见 §3.4 交互）。
* `market_regime` 必须**在参数分支前**可用（用于门控选择/提示）。
* `soft_window_alpha` 必须落盘/记录（监控用）。

---

#### 3.4.1 与参数分支（Strategy Params）

* **目的**：让参数分支在编码参数时**感知当前市场状态**与**上下文摘要**。
* **交互点**：

  * **早期注入**：参数分支的输入第一层前，拼接 `market_regime.onehot` 与 `context\_summary`（如 top‑k 统计、rv 等小向量，≤ 16 维）；
  * **MoE 门控**：MoE 的 gating（动态）以 `context_vec` 和 `market_regime` 为输入，输出 `G_t ∈ Δ^E`（E=专家数）。

#### 3.4.2 与排序/分类/回归头

* **concat 融合**：`[context_vec, params_vec, struct_embed] → scoring head`；
* **正则**（可选）：对 `context_vec` 使用 dropout（0.1–0.2）与 L2（1e-4）。

#### 3.4.3 与生成器/门控质检（可选）

* **条件生成**：生成器可读取 `market_summary`（与 `market_regime`）作为条件；
* **QC 指标**：可用 `soft_window_alpha` 分布、`rv_multi` 的分位偏差作为**质检输入**。

---

### 3.5 约束与非功能（NFR）

* **零泄漏**：任何特征/分位阈值/标准化统计**均在训练窗口内估计**，并在验证/测试**冻结**。
* **性能**：单周一次前向计算（不含训练）须 ≤ 10ms（CPU）/ ≤ 2ms（GPU），以 300 策略/周计。
* **可观测性**：记录 `α_w` 的均值/方差、`coverage_ratio`、`regime` 频次；输出 MLflow/日志。
* **版本化**：`feature_version`、`config hash`、`code sha` 均需写入 `meta`。
* **降级策略**：当 `coverage_ratio<0.6` 或 `α_w` 异常（NaN/Inf）时，回退到**单一默认窗口（30 天）**与**等权 α**，并上报警告。

---

### 4.1 数据与泄漏检验

* **AC-1**：给定 `as_of_date = 2024-03-31`，`market_window(30)` 的最晚日期必须 ≤ `2024-03-31`；
* **AC-2**：在任何 split（train/val/test）中，**参数估计（分位、标准化）**只使用 train 子集统计；val/test 仅**应用**。
* **AC-3**：窗口不足时 `mask=0` 生效，`coverage_ratio` 正确（单测：首 40 天，`W_set={60}` 时 `coverage_ratio≈2/3`）。

### 4.2 功能正确性

* **AC-4**：`soft_window_alpha` 对任一特征族满足 `α_w ≥ 0` 且 `∑α_w = 1`（误差 < 1e-6）。
* **AC-5**：`market_regime`（默认三分位法）在训练段的频率接近 1/3±10%；验证/测试仅应用阈值。
* **AC-6**：`context_vec` 维度一致（默认 128），NaN/Inf=0 且 `mask` 正确传递。

### 4.3 稳健性与回归测试

* **AC-7（漂移监控）**：新一批数据运行后，`α_w` 的分布与上一版本相比，KS 统计量 < 0.2，否则报警。
* **AC-8（容灾降级）**：模拟缺失 50% 的 OHLCV，系统自动降级到单窗口（30）+ 等权 α，并产出合法输出。
* **AC-9（性能）**：在 200 周 × 周均 300 的基线下，单次全量构造总时长 ≤ 1.5s（GPU）/ ≤ 8s（CPU）。

### 4.4 与下游的契约测试

* **AC-10（MoE 门控可用）**：在参数分支第一层前，能读到 `market_regime.onehot` 与 16 维 `context_summary`；
* **AC-11（排序头可用）**：拼接 `context_vec` 后，前向图正常，梯度可回传（e2e smoke test）。
* **AC-12（版本回放）**：同一 `feature_version/config sha` 下，重复运行输出一致（误差 < 1e-8）。

---

### 5.1 默认特征清单（features_v1）

* **价量**：`ret_1,5,10,20`; `rv_k`; `parkinson_k`; `rsi_14`; `macd(12,26,9)` 的 dif/dea/hist；`atr_k/close`; `vol_chg_k`; `corr_k(ret_1, volume)`；`amplitude_k`；`maxdrawdown_k`。
* **统计**：`mean_k, std_k, skew_k, kurt_k, q10_k, q25_k, q50_k, q75_k, q90_k, max/min 比, tail_ratio=(q90-q10)/q50`。
* **外生（可选）**：`funding_rate_mean_k`, `iv_level_k` 等。
* **尺度**：每个特征均在 W∈{7,14,21,30,60} 上计算并派生 `mask`。

### 5.2 软窗口实现要点

* 共享 MLP：`Input = [rv_30, ret_5, ret_20, q50_30, kurt_30]` 等**小摘要**；
* `α = softmax(Linear(ReLU(Linear(Input))))`；
* `f_multi = (α · f_W)`，逐特征族计算并拼接。

### 5.3 市场状态（默认三分位）

* 训练集上对 `rv_30` 取分位 `(q1≈33%, q2≈66%)`，定义 `Low=[0,q1)、Mid=[q1,q2)、High=[q2,∞)`；
* 推断期直接使用 `(q1, q2)` 阈值，不做再估计。

### 5.4 小型原始序列支路（可选）

* 输入：标准化后的 `close` 与 `volume`，长度 `max(W_set)`；
* 结构：`TCN(num_layers=2, channels=16, kernel_size=5) → GAP → FC(32)`；
* 正则：dropout=0.2、weight_decay=1e-4；默认 `enable_raw_branch=false`。

---

## 6. 配置与可调参数（默认值）

```
context_config_v1 = {
  "W_set": [7, 14, 21, 30, 60],
  "norm": {"method": "robust", "winsorize": {"p_low": 0.01, "p_high": 0.99}},
  "context_vec_dim": 128,
  "soft_window": {"hidden": 32, "share_mlp": true},
  "regime": {"method": "quantile", "states": 3, "feature": "rv_30"},
  "enable_raw_branch": false,
  "raw_branch": {"layers": 2, "channels": 16, "kernel": 5, "dropout": 0.2},
  "logging": {"record_alpha": true, "record_coverage": true}
}
```

---

## 7. 失败与降级场景（Fail & Fallback）

* **输入缺失或覆盖不足**：进入**降级模式**（单窗口 + 等权 α），并上报 `coverage_ratio`。
* **α 不可数（NaN/Inf）**：回退到等权 α，触发错误告警（但不中断流水线）。
* **regime 阈值文件缺失**：在线估计不可用时，**禁止**在线重估，使用**安全默认阈值**（训练期阈值的最后一次快照）。

---

## 8. 交付物与集成（Deliverables）

* **Python 模块/类**

  * `ContextFeatureBuilder.build(as_of_date, market_slice, config) -> outputs（§3.3）`
  * `RegimeDetector.fit(train_slice) / transform(slice)`
  * `SoftWindowAggregator.forward(feature_family_tensor) -> f_multi, alpha`
* **Schema/元数据**

  * `feature_version`, `config_sha`, `as_of_date`, `coverage_ratio`, `alpha_stats`
* **日志与监控**

  * MLflow：参数、度量（α 分布、regime 频次、PSI/KS）、工件（阈值、分位）。

---

## 9. 出厂自检清单（PM/开发/QA 三帽审阅）

* 开发能否仅凭本 FSD 实现 `build()` 并与下游对接？
* QA 能否据此写出数据泄漏、功能、稳健、性能与回归测试用例（§4 全覆盖）？
* 六个月后回看：**为何采用“工程特征 + 软窗口 + regime 门控”的归纳偏置？**（见 §1 原因：窄数据环境下的稳健性优先）

---

### 结语

本 FSD 把上下文分支做成**“可治理的黑盒组件”**：**确定的输入**、**严格的处理规则**、**契约化输出**、**可观测/可回放**。它既能在“最低算力”下稳健落地，也为后续**小型原始序列支路/图时空分支**保留了升级位，不会破坏现有契约。
